### PR TITLE
RATEST-294:New E2E: User Role Based Login

### PR DESCRIPTION
**Pull request by sherrif10**
https://issues.openmrs.org/browse/RATEST-294

### CODE REVIEW

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+@selenium
+@defaultLogin
+Scenario Outline: User Based login verification
+   When user verifies modules available on home page
+   And user verifies modules available on homepage after login as clerk
+   And user verifies modules available on home page after login as doctor
+   And user verifies modules available on home page after login as nurse
+   And user verifies modules available on home page after login as sysadmin
+   Then system goes back to login page
+
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:12]: are you able to remove the white spaces from the EOF line?

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+@selenium
+@defaultLogin
+Scenario Outline: User Based login verification
+   When user verifies modules available on home page
+   And user verifies modules available on homepage after login as clerk
+   And user verifies modules available on home page after login as doctor
+   And user verifies modules available on home page after login as nurse
+   And user verifies modules available on home page after login as sysadmin
+   Then system goes back to login page
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:11]: Add an empty line at the EOF to cater for git collaborators

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature (hunk 1 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature]: @kdaud  could you explain more of your statement above to make it more clear here.

**Comment by kdaud** [comment]
we can get rid of the [Login](https://github.com/openmrs/openmrs-contrib-qaframework/blob/master/qaframework-legacy-tests/src/test/java/org/openmrs/contrib/qaframework/legacy/LoginTest.java) selenium tests within this commit.

**Comment by sherrif10** [comment]
Nope thank you. LoginTest is vital. It s inherited to other classes. That's
why we named it differently.

**Comment by kdaud** [comment]
which classes inherit Login that will break when its deleted from the package?

**Comment by sherrif10** [comment]
Sorry my bad. Resolved.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature [hunk 3]
```diff
@@ -0,0 +1,11 @@
+Feature: User based login
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:1]:
```suggestion
Feature: User Based Login
```

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+@selenium
+@defaultLogin
+Scenario Outline: User Based login verification
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:5]: When do we use `Scenario Outline` over `Scenario` ?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature]: Probably we can use scenario since we got few outline tests.

**Review by kdaud — APPROVED** [review]
cc: @dkayiwa

**Comment by sherrif10** [comment]
Hello @dkayiwa  Requesting for reviewing on this guy.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java [hunk 5]
```diff
@@ -3,32 +3,35 @@
...
+import io.cucumber.java.Before;
+import io.cucumber.java.en.And;
+import io.cucumber.java.en.Then;
+import io.cucumber.java.en.When;
 
-public class LoginTest extends ReferenceApplicationTestBase {
+import org.openmrs.contrib.qaframework.RunTest;
 
-    private HomePage homePage;
+public class VerifyLoginSteps extends Steps {
```
Comment (dkayiwa) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:22]: Did you just forget to add a line here?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:22]: Resolved thank you.

**Comment by dkayiwa** [comment]
Shouldn't we also have some assertions for those that should not be present under less privileged accounts?

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/HomePage.java [hunk 6]
```diff
@@ -80,8 +80,14 @@ public Boolean isSystemAdministrationAppPresent() {
 	}
 
 	public Boolean isConfigureMetadataAppPresent() {
-		return isAppButtonPresent(CONFIGURE_METADATA_APP_ID);
+		try {
+			driver.findElement( CONFIGURE_METADATA);
```
Comment (dkayiwa) [inline qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/HomePage.java:84]: Why this change? Wasn't line 83 above enough?

Comment (sherrif10) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/HomePage.java]: Resolved it .

Comment (sherrif10) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/HomePage.java]: Resolved Thanks

**Comment by sherrif10** [comment]
Resolved this.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java [hunk 7]
```diff
@@ -39,43 +44,43 @@ public void verifyModulesAvailableOnHomePage() {
...
     }
 
-    @Test
-    @Category(BuildTests.class)
+    @And("user verifies modules available on home page after login as nurse")
     public void verifyNurseModulesAvailableOnHomePage() {
         goToLoginPage().loginAsNurse();
-        initiateHomePage();
         assertTrue(homePage.isFindAPatientAppPresent());
         assertTrue(homePage.isActiveVisitsAppPresent());
```
Comment (dkayiwa) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:69]: Is the nurse allowed to see the configure metadata app?

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java [hunk 8]
```diff
@@ -39,43 +44,43 @@ public void verifyModulesAvailableOnHomePage() {
...
 
-    @Test
-    @Category(BuildTests.class)
+    @And("user verifies modules available on home page after login as doctor")
     public void verifyDoctorModulesAvailableOnHomePage() {
         goToLoginPage().loginAsDoctor();
-        initiateHomePage();
         assertTrue(homePage.isFindAPatientAppPresent());
         assertTrue(homePage.isActiveVisitsAppPresent());
         assertTrue(homePage.isAppointmentSchedulingAppPresent());
```
Comment (dkayiwa) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:61]: Is the doctor allowed to see the systems administrator app?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:61]: All apps presented here are from homePage which the doctor is allowed to see.

Comment (dkayiwa) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:61]: Does the doctor have the system developer privileges?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:61]: No he doesnot.

Comment (dkayiwa) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:61]: Then why haven't you taken care of it in your assertions?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:61]: @dkayiwa  you can only view only these three apps after logging in as Doctor.

Comment (dkayiwa) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:61]: Is there anything in your test which proves that you cannot view the others?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:65]: No thanks.  Updated all the required scenarios.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    public void verifySysadminModulesAvailableOnHomePage() {
+        goToLoginPage().loginAsSysadmin();
+        assertTrue(homePage.isAppointmentSchedulingAppPresent());
+        assertTrue(homePage.isSystemAdministrationAppPresent());
+        assertFalse(homePage.isFindAPatientAppPresent());
+        assertFalse(homePage.isActiveVisitsAppPresent());
+        assertFalse(homePage.isCaptureVitalsAppPresent());
+        assertFalse(homePage.isRegisterPatientCustomizedForRefAppPresent());
+        assertFalse(homePage.isDataManagementAppPresent());
+        assertFalse(homePage.isConfigureMetadataAppPresent());
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:96]: is the system administrator less privileged to `DataManagementApp` and `ConfigureMetadataApp` ?

**Comment by kdaud** [comment]
> Ohh yes. try it with logging in with username = sysadmin , password=Sysadmin123

Have done the use case and confirmed that your correct!

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature [hunk 10]
```diff
@@ -0,0 +1,11 @@
+Feature: User Based Login
+
+@selenium
+@defaultLogin
+Scenario: User Based login verification
+   When user verifies modules available on home page
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:6]: the above verification is done by which user?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:6]: It is verified by default admin. which can be verified by ` username= admin, password=admin123`

Comment (kdaud) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:6]: how can one tell that from the gherkin step that is the admin(supper user) who is verifying ?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:6]: I knew this question will reflect :smile: . Ok

Comment (kdaud) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:6]: can one read the gherkin step and is able to know that the verification is done by the super user(admin)?

**Comment by sherrif10** [comment]
@dkayiwa  @kdaud  .At your convenient time, feel free to trigger the red button.

**Comment by kdaud** [comment]
@sherrif10 have you taken a look this [comment](https://github.com/openmrs/openmrs-contrib-qaframework/pull/369#discussion_r811619020)?

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature (hunk 10 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:6]: Thats is another answer given my broad explanation above. thats cool too.

**Comment by sherrif10** [comment]
Yes i did, does it need more action regarding the commit.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature (hunk 10 above)
Comment (kdaud) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:6]: can one read the gherkin step and is able to know that the verification is done by the super user(admin)?

**Comment by sherrif10** [comment]
There is also this patient too.

**Comment by sherrif10** [comment]
Like which ones @kdaud

**Comment by kdaud** [comment]
checkout at https://github.com/openmrs/openmrs-contrib-qaframework/pull/369#discussion_r811700007

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature (hunk 10 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:6]: Yes please.As long as your gherkin scenario is more readable and understandable and is doing the right thing

Comment (kdaud) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:6]: take a look at other steps and compare with the one above!

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:6]: @kdaud i dont see any problem, How would you rename that gherkin according to you.Any suggestion.

**Comment by sherrif10** [comment]
i answered your question above.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature (hunk 10 above)
Comment (kdaud) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:6]:
```js
When user verifies modules available on home page

```
compare with 
```js
And user verifies modules available on homepage after login as clerk
```

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:6]: > ```js
> login as clerk
> ```

Thanks its after login as super User

Comment (kdaud) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:6]: which user is logged in here?
```js
When user verifies modules available on home page
```

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature]: you need to probably re-read my comment above that explains the workflow and functionalities of the first scenario. this is called a super user . Have updated the scenario to make it more meaning.

**Comment by kdaud** [comment]
@sherrif10 I see almost all the checks are red! Are you looking into them!

**Comment by sherrif10** [comment]
Am also wondering why they broke but am looking into it.

**Comment by sherrif10** [comment]
@kdaud  seems we are having issues with ci buid.

**Comment by kdaud** [comment]
> @kdaud seems we are having issues with ci buid.

are you investigating the issues with ci build?

**Comment by sherrif10** [comment]
Sure i can see you doing the same.

**Comment by sherrif10** [comment]
There is a problem with our server here https://qa-refapp.openmrs.org/openmrs/error.html

**Comment by kdaud** [comment]
how can I reproduce the issue?

**Comment by sherrif10** [comment]
@kdaud  Have noticed that at some level of point when our qa server is on over load. we ought to give it time and re run our workflows to pass Like this very case.So they are happily passing now.

**Comment by kdaud** [comment]
@sherrif10 are you adding `CheckModuleTest` in this workflow?

**Comment by sherrif10** [comment]
Have added it there.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java (hunk 7 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:84]: Nurse is not allowed to see configure metadata app.

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/HomePage.java [hunk 11]
```diff
@@ -106,6 +107,7 @@ public void goToManageForm() {
 	}
 
 	public AdministrationPage goToAdministration() {
+		waiter.until(ExpectedConditions.visibilityOfElementLocated(SYSTEM_ADMINISTRATION));
```
Comment (kdaud) [inline qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/HomePage.java:110]: any reason for adding the waiter call here?

File: .github/workflows/platform-install-postgres.yml [hunk 12]
```diff
@@ -3,7 +3,7 @@ on:
   push:
     branches: [ master ]
   repository_dispatch:
-    types: [qa-initialSetup]
+    types: [qa]
```
Comment (kdaud) [inline .github/workflows/platform-install-postgres.yml:6]: ensure it's only your changes that are being reflected?

Comment (sherrif10) [reply .github/workflows/platform-install-postgres.yml]: Thanks. i resolved that.

**Comment by sherrif10** [comment]
@kdaud Does this look good to be merged.

**Comment by kdaud** [comment]
@sherrif10 did you just forget to get rid of [CheckModuleTest](https://github.com/openmrs/openmrs-contrib-qaframework/blob/d1af86de161d5e915e659c1fa351450f08e85198/qaframework-legacy-tests/src/test/java/org/openmrs/contrib/qaframework/legacy/CheckModulesTest.java#L24)?

**Comment by sherrif10** [comment]
Opps my git reset --hard overriden that. Fixed it asap.

**Review by kdaud — APPROVED** [review]
cc: @dkayiwa

**Comment by sherrif10** [comment]
Hello @dkayiwa We kindly request for approval.

File: qaframework-legacy-tests/src/test/java/org/openmrs/contrib/qaframework/legacy/CheckModulesTest.java [hunk 13]
```diff
@@ -1,44 +0,0 @@
...
- */
-package org.openmrs.contrib.qaframework.legacy;
-
-import static org.junit.Assert.assertFalse;
-
-import java.util.List;
-
-import org.junit.Test;
-import org.junit.experimental.categories.Category;
-import org.openmrs.contrib.qaframework.helper.BuildTests;
```
Comment (dkayiwa) [inline qaframework-legacy-tests/src/test/java/org/openmrs/contrib/qaframework/legacy/CheckModulesTest.java:18]: Are these deletions related to the ticket at hand?

Comment (sherrif10) [reply qaframework-legacy-tests/src/test/java/org/openmrs/contrib/qaframework/legacy/CheckModulesTest.java:18]: Yes they are part of the test we are getting rid of the whole class.

Comment (dkayiwa) [reply qaframework-legacy-tests/src/test/java/org/openmrs/contrib/qaframework/legacy/CheckModulesTest.java:18]: Can you explain how it is related to User Role Based Login?

Comment (sherrif10) [reply qaframework-legacy-tests/src/test/java/org/openmrs/contrib/qaframework/legacy/CheckModulesTest.java:18]: It is not related to user Role Based Login , However since it checks which module gets started/Not started  with in the server, it would be good deal to also include it here, However am of no objection to remove it. cc @kdaud  do you have suggestion on this.

Comment (dkayiwa) [reply qaframework-legacy-tests/src/test/java/org/openmrs/contrib/qaframework/legacy/CheckModulesTest.java:18]: Do you mind taking a look at this? https://opensource.com/article/18/6/anatomy-perfect-pull-request

Comment (sherrif10) [reply qaframework-legacy-tests/src/test/java/org/openmrs/contrib/qaframework/legacy/CheckModulesTest.java:18]: Sure thanks. Let me correct it .

**Comment by sherrif10** [comment]
Hey  @dkayiwa :smiley: . Am patiently waiting on this. If its ready i think we can merge them.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+        initiateWithLogin();
+    }
+    
+    @After(RunTest.HOOK.SELENIUM_DEFAULT_LOGIN)
+    public void destroy() {
+        quit();
+    }
+
+    @And("user verifies modules available on home page after login as super user")
+    public void verifyModulesAvailableOnHomePage() {
```
Comment (dkayiwa) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:34]: From the method name, how do i tell the user role?

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java (hunk 9 above)
Comment (dkayiwa) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:94]: @sherrif10 did you address the above comment from @kdaud ?

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java (hunk 14 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:34]: Thanks, @dkayiwa for reviewing again. You won't tell the role from the method name. However the method that triggers the functionality gets initiated with `initiateLoginPage`.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java (hunk 9 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:94]: Yes @kdaud  system administration is less privileged to the two apps.

Comment (dkayiwa) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:94]: @sherrif10 how can that be?

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java (hunk 14 above)
Comment (dkayiwa) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:34]: What is the difference between `super user` and `sys admin`?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:34]: We have two login crendentials for the two. Super user logins in as admin, admin123 which is our default and we inherit them from https://github.com/openmrs/openmrs-contrib-qaframework/blob/master/qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/Steps.java#L71 basically the one we use for test.Properties
then for sys admin. he is entitled for login crendentials as sysadmin and sysadmin123 as seen here https://github.com/openmrs/openmrs-contrib-qaframework/blob/master/qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/helper/LoginPage.java#L41

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java (hunk 9 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:94]: For confirm this. We can use  "sysadmin";  and  "Sysadmin123";

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java (hunk 14 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:34]: cc @dkayiwa .

Comment (dkayiwa) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:34]: I find it confusing to have a sys admin without full privileges.

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:34]: Sys admin have full privileges. Its true and all apps can be accessed unless am missing out some apps that are accessed. For those in systemAdministration, we wont be able to access them via home page but instead systemadmin page

Comment (dkayiwa) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:34]: If they are not on the home page, then why do we have them in these assertions? I thought your tests are about what buttons different roles should see on the home page.

**Comment by sherrif10** [comment]
cc @dkayiwa

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java (hunk 14 above)
Comment (dkayiwa) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:34]: @sherrif10 did you see the above comment?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:34]: ohh. sorry. For the comment above. Yes  tests are about what buttons different roles should see on the home page. we only have those assertion to false for those not accessible on home page and those assertion to true those accessible on home page. forexample if user logs in using sysadmin and sysAdmin123, you can only see two apps. on that due not , am using assertTrue( for the two apps). others not visible are false.

Comment (dkayiwa) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:34]: Then why do we have assertions for buttons that do not appear on the home page?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:34]: @dkayiwa  Different login credentials have different apps appearing on the home page.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java [hunk 15]
```diff
@@ -40,6 +40,9 @@ public void verifyModulesAvailableOnHomePage() {
         assertTrue(homePage.isDataManagementAppPresent());
         assertTrue(homePage.isConfigureMetadataAppPresent());
         assertTrue(homePage.isSystemAdministrationAppPresent());
+        assertTrue(homePage.isDataManagementAppPresent());
+        
+
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:45]: any reason for adding the empty lines here?

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java (hunk 14 above)
Comment (dkayiwa) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:34]: But i see you asserting for apps that will never appear on the home page regardless of who is logged in.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java [hunk 16]
```diff
@@ -3,32 +3,35 @@
...
-        assertPage(homePage.waitForPage());
+    @After(RunTest.HOOK.SELENIUM_DEFAULT_LOGIN)
+    public void destroy() {
+        quit();
     }
 
-    @Test
-    @Category(BuildTests.class)
+    @And("user verifies modules available on home page after login as admin")
     public void verifyModulesAvailableOnHomePage() {
```
Comment (dkayiwa) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:34]: This method name needs to have the role that it is testing

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java [hunk 17]
```diff
@@ -39,43 +42,42 @@ public void verifyModulesAvailableOnHomePage() {
...
-        initiateHomePage();
         assertTrue(homePage.isFindAPatientAppPresent());
         assertTrue(homePage.isActiveVisitsAppPresent());
         assertTrue(homePage.isAppointmentSchedulingAppPresent());
         assertTrue(homePage.isCaptureVitalsAppPresent());
     }
 
-    @Test
-    @Category(BuildTests.class)
+    @And("user verifies modules available on home page after login as sysadmin")
```
Comment (dkayiwa) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:70]: How is login as admin different from login as sysadmin?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:70]: i think the confusion is with in login page where this was made. admin like you had said is the super user. then for this case we have sysadmin and SysAdmin123. i think these are two different cases from where they differ.

Comment (dkayiwa) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:70]: Anything that has `admin` is super user.  Else, you are creating unnecessary confusion.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java (hunk 16 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:34]: I think i will change method name.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java (hunk 17 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:70]: I get your point and where the confusion is. The thing is this is right but the naming of gherkin is bringing confusion, the difference is sysadmin can only access to the system administration app and appointment scheduling but admin can access all the apps.Does this make sense.

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:70]: can we however change the naming of gherkin

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:70]: @dkayiwa does it now add a difference.

Comment (dkayiwa) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:70]: It will never make sense as long as an admin and sysadmin are different.

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:70]: @dkayiwa  How about  this [commit](https://github.com/openmrs/openmrs-contrib-qaframework/pull/369/commits/ff309d05402b6f9d3e9bee298363b63785259817) where by i added login as admin to differentiate it from loginAsSysadmin.

Comment (dkayiwa) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:70]: admin, sysadmin, and super user are all the same. Else, you are creating confusing tests.

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:70]: Ohh so are you suggesting we remove the one for sysadmin.

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java]: i removed the one for sysadmin. because it kind of duplicate.

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java]: @dkayiwa  hope this looks fair.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java (hunk 15 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java]: Resolved

**Comment by dkayiwa** [comment]
@sherrif10 you still have unresolved comments.

**Comment by sherrif10** [comment]
@dkayiwa  i  guess i addressed all of them.

**Comment by dkayiwa** [comment]
Can you confirm?

**Comment by sherrif10** [comment]
Its unfortunate that i cant see any pending review. Could you kindly point that to me.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java (hunk 16 above)
Comment (dkayiwa) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:33]: Did you address the above?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java]: Resolved. Thanks.

**Comment by dkayiwa** [comment]
How do i tell that buttons which i do not expect to show up basing on certain roles, are not present?

**Comment by sherrif10** [comment]
> How do i tell that buttons which i do not expect to show up basing on certain roles, are not present?

We can add those as `AssertFalse(homepage.isNameOfTheAppPresent)`, but had rejected this.

**Comment by dkayiwa** [comment]
> but had rejected this.

Who rejected this?

**Comment by sherrif10** [comment]
Added assert false for those apps which we dont expect to appear given user role privilege respectively.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+    @After(RunTest.HOOK.SELENIUM_DEFAULT_LOGIN)
+    public void destroy() {
+        quit();
+    }
+
+    @And("user verifies modules available on home page after login as admin")
+    public void verifyAdminModulesAvailableOnHomePage() {
+        goToLoginPage().loginAsAdmin();
+        assertTrue(homePage.isFindAPatientAppPresent());
+        assertTrue(homePage.isActiveVisitsAppPresent());
```
Comment (dkayiwa) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:37]: I see none buttons here: https://qa-refapp.openmrs.org/openmrs/referenceapplication/home.page
Is there any reason why this guy can only see eight?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:37]: We didnot consider report app since it doesnt have much work , we also didnot include it on home page. Unless you are okay me adding it here.
